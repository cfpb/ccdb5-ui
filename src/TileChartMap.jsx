import './TileChartMap.less'
import { connect } from 'react-redux'
import { hashObject } from './utils'
import React from 'react'
import { TileMap } from 'cfpb-chart-builder'

export class TileChartMap extends React.Component {
  componentDidUpdate( prevProps ) {
    const props = this.props
    if ( !props.data[0].length ) {
      return
    }

    if ( hashObject( prevProps.data ) !== hashObject( props.data ) ) {
      this._redrawMap()
    }
  }

  render() {
    return (
      <div>
        <div id="tile-chart-map"
             className="cfpb-chart"
             data-chart-type="tile_map">
        </div>
      </div>
    )
  }

  // --------------------------------------------------------------------------
  // Event Handlers
  _redrawMap() {
    const colors = [
      'rgba(247, 248, 249, 0.5)',
      'rgba(212, 231, 230, 0.5)',
      'rgba(180, 210, 209, 0.5)',
      'rgba(137, 182, 181, 0.5)',
      'rgba(86, 149, 148, 0.5)',
      'rgba(37, 116, 115, 0.5)'
    ]

    // eslint-disable-next-line no-unused-vars
    const chart = new TileMap( {
      el: document.getElementById( 'tile-chart-map' ),
      data: this.props.data,
      colors,
      localize: true
    } )
  }
}

export const mapStateToProps = state => ( { data: [ state.map.state ]} )

export default connect( mapStateToProps )( TileChartMap )

