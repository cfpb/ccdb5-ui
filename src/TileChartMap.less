/**
 * @license Highcharts
 *
 * (c) 2009-2016 Torstein Honsi
 *
 * License: www.highcharts.com/license
 */
.highcharts-container {
    position: relative;
    overflow: hidden;
    width: 100%;
    height: 100%;
    text-align: left;
    line-height: normal;
    z-index: 0;
    /* #1072 */
    -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
    font-family: "Lucida Grande", "Lucida Sans Unicode", Arial, Helvetica, sans-serif;
    font-size: 12px;
}
.highcharts-root {
    display: block;
    text {
        stroke-width: 0;
    }
}

.highcharts-strong {
    font-weight: bold;
}
.highcharts-emphasized {
    font-style: italic;
}
.highcharts-anchor {
    cursor: pointer;
}
.highcharts-background {
    fill: #ffffff;
}
.highcharts-plot-border,
.highcharts-plot-background {
    fill: none;
}
.highcharts-label-box {
    fill: none;
}
.highcharts-button-box {
    fill: inherit;
}
/* Titles */
.highcharts-title {
    fill: #333333;
    font-size: 1.5em;
}
.highcharts-subtitle {
    fill: #666666;
}

/* Highmaps */
.highcharts-map-series {
    .highcharts-point {
        -webkit-transition: fill 500ms, fill-opacity 500ms, stroke-width 250ms;
        transition: fill 500ms, fill-opacity 500ms, stroke-width 250ms;
        stroke: #cccccc;
    }

    .highcharts-point-hover {
        -webkit-transition: fill 0ms, fill-opacity 0ms;
        transition: fill 0ms, fill-opacity 0ms;
        fill-opacity: 0.5;
        stroke-width: 2px;
    }
}

@media only all and (max-width: 37.5em) {
}@media only all and (max-width: 37.5em) {
}/*
Updating the charts:
1. Set the positioning of the chart UI elements using the JavaScript API.
2. Add in styles below that aren't possible through the API.
This appears to handle responsive changes the best, since the JS sets
x/y positioning on the SVG elements.
*/
.u-chart-label {
    font-size: 16px;
    font-family: inherit;
    color: #5a5d61;
}

.cfpb-chart {
    position: relative;
    max-width: 650px;
    min-width: 320px;
    height: 500px;
    margin-bottom: 25px;
    .highcharts-root,
    .highcharts-container,
    .highcharts-legend-box,
    .highcharts-axis-labels,
    .highcharts-tooltip text {
        font-size: 16px;
        font-family: inherit;
        color: #5a5d61;
    }
}

/* TODO: Investigate whether responsive styles can be remove here and handled
         via class set in JS, like how the other chart types work. */
.cfpb-chart[data-chart-type="tile_map"] {
    .highcharts-label {
        pointer-events: none;
    }
    .highcharts-legend__tile-map {
        font-size: 12px;
    }
    .highcharts-label__tile-map {
        font-size: 16px;
        font-family: inherit;
        color: #5a5d61;
        color: #101820;
        font-weight: 600;

        span {
            position: relative !important;
            white-space: normal !important;
        }
    }

    .u-point-hover-state {
        stroke: #101820;
        cursor: default;
        stroke-width: 4px !important;
        stroke-linecap: square;
    }
    .highcharts-map-series {
        .highcharts-point {
            stroke: #75787b;
        }

        .highcharts-point:focus, .highcharts-point:hover {
            stroke: #101820;
            cursor: default;
            stroke-width: 4px !important;
            stroke-linecap: square;
        }

        .highcharts-point-hover {
            stroke: #101820;
            cursor: default;
            stroke-width: 4px !important;
            stroke-linecap: square;
        }
    }
    .highcharts-data-label {
        font-weight: 600;
        color: #101820;
        font-size: 14px;
        line-height: 1.3;

        .highcharts-data-label-state {
            text-align: center;
            cursor: default;

            .abbr {
                font-size: 16px;
            }

            .value {
                font-size: 12px;
            }
        }
    }

    .highcharts-label-box {
        display: none;
    }

    .highcharts-tooltip {
        min-width: 200px;

        span {
            color: #fff;
            background-color: #101820;
            padding: 5px;
            width: 320px;

            div p {
                margin: 5px;
                white-space: normal;

                &.u-float-left {
                    width: 150px;
                    float: left;
                }

                &.u-right {
                    width: 150px;
                }
            }

            .title {
                text-align: center;
            }

            .row {
                border-top: solid 1px #43484e;
                margin-top: 5px;
                padding-top: 5px;
            }
        }
    }
}
@media only all and (max-width: 37.5em) {
    .cfpb-chart[data-chart-type="tile_map"] .highcharts-data-label {
        margin-top: 2px;
        font-size: 10px;
        line-height: 1;
    }
}

@media only all and (max-width: 37.5em) {
    .cfpb-chart[data-chart-type="tile_map"] .highcharts-data-label .highcharts-data-label-state .value {
        display: block;
        margin-top: -2px;
        font-size: 9px;
    }
}

@media (max-width: 600px) {
    .cfpb-chart[data-chart-type="tile_map"] {
        height: 400px;
    }
}
.no-js .cfpb-chart {
    height: inherit;
}
.no-js .cfpb-chart:after {
    border: 0;
}
