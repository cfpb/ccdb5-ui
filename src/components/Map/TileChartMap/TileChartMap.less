/**
 * @license Highcharts
 *
 * (c) 2009-2016 Torstein Honsi
 *
 * License: www.highcharts.com/license
 */
.highcharts-container {
  position: relative;
  overflow: hidden;
  width: 100%;
  height: 100%;
  text-align: left;
  line-height: normal;
  z-index: 0;

  /* #1072 */
  -webkit-tap-highlight-color: rgba(0, 0, 0, 0%);
  font-family: 'Lucida Grande', 'Lucida Sans Unicode', Arial, Helvetica,
    sans-serif;
  font-size: 12px;
}
// legend
.highcharts-root {
  display: block;
  text {
    stroke-width: 0;
    font-weight: 500;
  }
}

.highcharts-strong {
  font-weight: bold;
}

.highcharts-emphasized {
  font-style: italic;
}

.highcharts-anchor {
  cursor: pointer;
}

.highcharts-background {
  fill: @white;
}

.highcharts-plot-border,
.highcharts-plot-background {
  fill: none;
}

.highcharts-label-box {
  fill: none;
}

.highcharts-button-box {
  fill: inherit;
}

/* Titles */
.highcharts-title {
  fill: #333;
  font-size: 1.5em;
}

.highcharts-subtitle {
  fill: #666;
}

/* Highmaps */
.highcharts-map-series {
  .highcharts-point {
    transition: fill 500ms, fill-opacity 500ms, stroke-width 250ms;
    stroke: #ccc;
  }

  .highcharts-point-hover {
    transition: fill 0ms, fill-opacity 0ms;
    cursor: default;
    fill: @pacific !important;
    fill-opacity: 0.5;
  }
}

/*
Updating the charts:
1. Set the positioning of the chart UI elements using the JavaScript API.
2. Add in styles below that aren't possible through the API.
This appears to handle responsive changes the best, since the JS sets
x/y positioning on the SVG elements.
*/
.u-chart-label {
  font-size: 16px;
  font-family: inherit;
  color: @gray;
}

.cfpb-chart {
  position: relative;
  width: ~'calc(100% - 60px)';
  min-height: 400px;
  margin-bottom: 25px;
  margin-left: @grid_gutter-width;
  .highcharts-root,
  .highcharts-container,
  .highcharts-legend-box,
  .highcharts-axis-labels,
  .highcharts-tooltip text {
    font-size: 16px;
    font-family: inherit;
    color: @gray;
  }
}

/* TODO: Investigate whether responsive styles can be remove here and handled
         via class set in JS, like how the other chart types work. */
.cfpb-chart[data-chart-type='tile_map'] {
  .highcharts-data-label {
    font-weight: 500;
    color: @black;

    .highcharts-data-label-state {
      text-align: center;
      cursor: default;
      &.deselected {
        color: @gray;
      }
    }
  }

  .highcharts-label {
    pointer-events: none;
  }

  .highcharts-legend-title {
    .highcharts-legend-key {
      font-weight: 500;
    }
    .highcharts-legend-description {
      font-size: 12px;
      .type {
        font-weight: 500;
        color: @black;
      }
    }
  }

  .highcharts-legend__tile-map {
    font-size: 12px;

    .legend-box {
      stroke: @gray;
    }
  }

  .highcharts-label__tile-map {
    font-size: 16px;
    font-family: inherit;
    color: @gray;
    font-weight: 500;

    span {
      position: relative !important;
      white-space: normal !important;
    }
  }

  .u-point-hover-state,
  .selected {
    stroke: @black;
    cursor: default;
    stroke-linecap: square;
  }
  .highcharts-map-series {
    .highcharts-point {
      stroke: none;

      &.empty {
        stroke: rgba(231, 232, 233, 100%);
      }

      &.selected {
        stroke: rgba(0, 45, 114, 100%);
      }
    }
  }

  .highcharts-label-box {
    display: none;
  }

  .highcharts-tooltip {
    min-width: 200px;

    span {
      color: @white;
      background-color: @black;
      padding: 5px;
      width: 320px;

      div p {
        margin: 5px;
        white-space: normal;

        &.u-float-left {
          width: 150px;
          float: left;
        }

        &.u-right {
          width: 150px;
        }
      }

      .title {
        text-align: center;
      }

      .row {
        border-top: solid 1px #43484e;
        margin-top: 5px;
        padding-top: 5px;
      }
    }
  }
}

.cfpb-chart[data-chart-type='tile_map']:not(.print) {
  .highcharts-map-series {
    .highcharts-point {
      &:hover {
        stroke: rgba(0, 45, 114, 100%);
        stroke-width: 2px !important;
      }
    }
  }
}

@media @tablet-sm {
  .cfpb-chart[data-chart-type='tile_map'] {
    width: ~'calc(100% - 20px)';
    margin-bottom: 10px;
    margin-left: 10px;
  }
}

@media @phone {
  .cfpb-chart[data-chart-type='tile_map'] {
    width: ~'calc(100% - 20px)';
    margin-bottom: 10px;
    margin-left: 10px;
  }
}

// -----------------------------------------------------------------------------
// State Tile Break Points

@mobile-tiles: ~'only screen and (max-width:1024px)';
@mobile-xs-tiles: ~'only screen and (max-width:400px)';

.highcharts-data-label {
  font-size: 14px;
  line-height: 1.3;

  .highcharts-data-label-state {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    // IE 11 fixes
    @media all and (-ms-high-contrast: none), (-ms-high-contrast: active) {
      display: inline-block;
      float: left;
    }

    .abbr {
      flex: 1 0 auto;
      font-size: 16px;
    }

    .value {
      flex: 1 0 auto;
      margin-top: 6px;
      font-size: 12px;
    }
    &.hover {
      .abbr,
      .value {
        text-decoration: underline;
      }
    }
    &.deselected {
      color: rgba(90, 93, 97, 100%);
    }
  }
}

// This takes over for looking good at the minimum height of 400px
@media @mobile-tiles {
  .highcharts-data-label {
    line-height: 1;

    .highcharts-data-label-state {
      .abbr {
        font-size: 12px;
      }
      .value {
        margin-top: 3px;
        font-size: 9px;
      }
    }
  }
}

@media @mobile-xs-tiles {
  .highcharts-data-label {
    line-height: 1;

    .highcharts-data-label-state {
      // IE 11 shift tile labels up
      @media (-ms-high-contrast: none), (-ms-high-contrast: active) {
        transform: translateY(-10px);
      }
      .abbr {
        font-size: 12px;
      }
      .value {
        display: none;
      }
    }
  }
}

// -----------------------------------------------------------------------------
// Make the map & legend play nice

@media @tablet-sm {
  .highcharts-series-group {
    transform: translate(0, 15px);
  }

  .highcharts-data-labels.highcharts-map-series {
    margin-top: 15px;
  }

  .highcharts-data-label-state {
    // IE 11 shift tile labels up
    @media (-ms-high-contrast: none), (-ms-high-contrast: active) {
      transform: translateY(-10px);
    }
  }
}

@media @phone {
  .highcharts-series-group {
    transform: translate(0, 15px);
  }

  .highcharts-data-labels.highcharts-map-series {
    margin-top: 15px;
  }
}

.no-js .cfpb-chart {
  height: inherit;
}

.no-js .cfpb-chart::after {
  border: 0;
}
