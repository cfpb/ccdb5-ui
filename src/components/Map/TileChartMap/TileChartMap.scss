/**
 * @license Highcharts
 *
 * (c) 2009-2016 Torstein Honsi
 *
 * License: www.highcharts.com/license
 */
@use '@cfpb/cfpb-design-system/src/abstracts' as *;
@use 'css/base' as *;

.highcharts-container {
  position: relative;
  overflow: hidden;
  width: 100%;
  height: 100%;
  text-align: left;
  line-height: normal;
  z-index: 0;

  /* #1072 */
  -webkit-tap-highlight-color: rgb(0 0 0 / 0%);
  font-family: 'Lucida Grande', 'Lucida Sans Unicode', Arial, Helvetica,
    sans-serif;
  font-size: 12px;
}

// legend
.highcharts-root {
  display: block;
  text {
    stroke-width: 0;
    font-weight: 500;
  }
}

.highcharts-strong {
  font-weight: bold;
}

.highcharts-emphasized {
  font-style: italic;
}

.highcharts-anchor {
  cursor: pointer;
}

.highcharts-background {
  fill: var(--white);
}

.highcharts-plot-border,
.highcharts-plot-background {
  fill: none;
}

.highcharts-label-box {
  fill: none;
}

.highcharts-button-box {
  fill: inherit;
}

/* Titles */
.highcharts-title {
  fill: #333;
  font-size: 1.5em;
}

.highcharts-subtitle {
  fill: #666;
}

/* Highmaps */
.highcharts-map-series {
  .highcharts-point {
    transition:
      fill 500ms,
      fill-opacity 500ms,
      stroke-width 250ms;
    stroke: #ccc;
  }

  .highcharts-point-hover {
    transition:
      fill 0ms,
      fill-opacity 0ms;
    cursor: default;
    fill: var(--pacific) !important;
    fill-opacity: 0.5;
  }
}

/*
Updating the charts:
1. Set the positioning of the chart UI elements using the JavaScript API.
2. Add in styles below that aren't possible through the API.
This appears to handle responsive changes the best, since the JS sets
x/y positioning on the SVG elements.
*/
.u-chart-label {
  font-size: 16px;
  font-family: inherit;
  color: var(--gray);
}

.cfpb-chart {
  position: relative;
  width: calc(100% - 60px);
  min-height: 400px;
  margin-bottom: 25px;
  margin-left: $grid-gutter-width;
  .highcharts-root,
  .highcharts-container,
  .highcharts-legend-box,
  .highcharts-axis-labels,
  .highcharts-tooltip text {
    font-size: 16px;
    font-family: inherit;
    color: var(--gray);
  }
}

/* TODO: Investigate whether responsive styles can be remove here and handled
         via class set in JS, like how the other chart types work. */
.cfpb-chart[data-chart-type='tile_map'] {
  .highcharts-data-label {
    font-weight: 500;
    color: var(--black);

    .highcharts-data-label-state {
      text-align: center;
      cursor: default;
      &.deselected {
        color: var(--gray);
      }
    }
  }

  .highcharts-label {
    pointer-events: none;
  }

  .highcharts-legend-title {
    .highcharts-legend-key {
      font-weight: 500;
    }
    .highcharts-legend-description {
      font-size: 12px;
      .type {
        font-weight: 500;
        color: var(--black);
      }
    }
  }

  .highcharts-legend__tile-map {
    font-size: 12px;

    .legend-box {
      stroke: var(--gray);
    }
  }

  .highcharts-label__tile-map {
    font-size: 16px;
    font-family: inherit;
    color: var(--gray);
    font-weight: 500;

    span {
      position: relative !important;
      white-space: normal !important;
    }
  }

  .u-point-hover-state,
  .selected {
    stroke: var(--black);
    cursor: default;
    stroke-linecap: square;
  }
  .highcharts-map-series {
    .highcharts-point {
      stroke: none;

      &.empty {
        stroke: rgb(231 232 233 / 100%);
      }

      &.selected {
        stroke: rgb(0 45 114 / 100%);
      }
    }
  }

  .highcharts-label-box {
    display: none;
  }

  .highcharts-tooltip {
    min-width: 200px;

    span {
      color: var(--white);
      background-color: var(--black);
      padding: 5px;
      width: 320px;

      div p {
        margin: 5px;
        white-space: normal;

        &.u-float-left {
          width: 150px;
          float: left;
        }

        &.u-right {
          width: 150px;
        }
      }

      .title {
        text-align: center;
      }

      .row {
        border-top: solid 1px #43484e;
        margin-top: 5px;
        padding-top: 5px;
      }
    }
  }
}

.cfpb-chart[data-chart-type='tile_map']:not(.print) {
  .highcharts-map-series {
    .highcharts-point {
      &:hover {
        stroke: rgb(0 45 114 / 100%);
        stroke-width: 2px !important;
      }
    }
  }
}

@media screen and ((min-width: $layout-breakpoint-tablet-sm-min) and (max-width: $layout-breakpoint-tablet-sm-max)),
screen and (max-width: $layout-breakpoint-phone-max) {
  .cfpb-chart[data-chart-type='tile_map'] {
    width: calc(100% - 20px);
    margin-bottom: 10px;
    margin-left: 10px;
  }
}


// -----------------------------------------------------------------------------
// State Tile Break Points
$mobile-tiles-breakpoint-max: 1024px;
$mobile-xs-tiles-breakpoint-max: 400px;

.highcharts-data-label {
  font-size: 14px;
  line-height: 1.3;

  .highcharts-data-label-state {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;

    .abbr {
      flex: 1 0 auto;
      font-size: 16px;
    }

    .value {
      flex: 1 0 auto;
      margin-top: 6px;
      font-size: 12px;
    }
    &.hover {
      .abbr,
      .value {
        text-decoration: underline;
      }
    }
    &.deselected {
      color: rgb(90 93 97 / 100%);
    }
  }
}

// This takes over for looking good at the minimum height of 400px
@media screen and (max-width: $mobile-tiles-breakpoint-max) {
  .highcharts-data-label {
    line-height: 1;

    .highcharts-data-label-state {
      .abbr {
        font-size: 12px;
      }
      .value {
        margin-top: 3px;
        font-size: 9px;
      }
    }
  }
}

@media screen and (max-width: $mobile-xs-tiles-breakpoint-max) {
  .highcharts-data-label {
    line-height: 1;

    .highcharts-data-label-state {
      .abbr {
        font-size: 12px;
      }
      .value {
        display: none;
      }
    }
  }
}

// -----------------------------------------------------------------------------
// Make the map & legend play nice

@media screen and (min-width: $layout-breakpoint-tablet-sm-min) and (max-width: $layout-breakpoint-tablet-sm-max) {
  .highcharts-series-group {
    transform: translate(0, 15px);
  }

  .highcharts-data-labels.highcharts-map-series {
    margin-top: 15px;
  }
}

@media screen and (max-width: $layout-breakpoint-phone-max) {
  .highcharts-series-group {
    transform: translate(0, 15px);
  }

  .highcharts-data-labels.highcharts-map-series {
    margin-top: 15px;
  }
}

.no-js .cfpb-chart {
  height: inherit;
}

.no-js .cfpb-chart::after {
  border: 0;
}
