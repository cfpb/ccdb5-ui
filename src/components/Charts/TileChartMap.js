import './TileChartMap.less';
import { addStateFilter, removeStateFilter } from '../../actions/map';
import { coalesce, hashObject, sendAnalyticsEvent } from '../../utils';
import { GEO_NORM_NONE, STATE_DATA } from '../../constants';
import { connect } from 'react-redux';
import React from 'react';
import TileMap from './TileMap';

export class TileChartMap extends React.Component {
  componentDidUpdate(prevProps) {
    const props = this.props;
    if (!props.data[0].length) {
      return;
    }

    // force redraw when switching tabs, or print mode changes
    if (
      hashObject(prevProps) !== hashObject(props) ||
      !document.getElementById('tile-chart-map').children.length
    ) {
      this._redrawMap();
    }
  }

  render() {
    return (
      <div>
        <div
          id="tile-chart-map"
          className={'cfpb-chart ' + this.props.printClass}
          data-chart-type="tile_map"
        ></div>
      </div>
    );
  }

  _toggleState(event) {
    // pass in redux dispatch
    // point.fullName
    const compProps = this;
    const { abbr, fullName } = event.point;
    const selectedState = {
      abbr,
      // rename this for consistency
      // chart builder uses fullName
      name: fullName,
    };
    if (compProps.stateFilters.includes(abbr)) {
      compProps.removeState(selectedState);
    } else {
      compProps.addState(selectedState);
    }
  }

  // --------------------------------------------------------------------------
  // Event Handlers

  _redrawMap() {
    const toggleState = this._toggleState;
    const componentProps = this.props;
    const mapElement = document.getElementById('tile-chart-map');
    const { dataNormalization, hasTip } = componentProps;
    const width = mapElement.clientWidth || 700;
    const data = updateData(this.props);

    const options = {
      el: mapElement,
      data,
      isPerCapita: dataNormalization !== GEO_NORM_NONE,
      events: {
        // custom event handlers we can pass on
        click: toggleState.bind(componentProps),
      },
      hasTip,
      width,
    };

    options.height = width * 0.75;

    // eslint-disable-next-line no-unused-vars
    const chart = new TileMap(options);
  }
}

/**
 * helper function to get display value of tile based on selected dropdown.
 * @param {object} props contains data and normalization
 * @returns {object} data provided to tile map
 */
function updateData(props) {
  const { data, dataNormalization } = props;
  const showDefault = dataNormalization === GEO_NORM_NONE;
  const res = data[0].map((o) => ({
    ...o,
    displayValue: showDefault ? o.value : o.perCapita,
  }));

  return res;
}

/**
 * helper function to calculate percapita value
 * @param {object} stateObj a state containing abbr and value
 * @param {object} stateInfo other information about the state
 * @returns {string} the Per 1000 population value
 */
function getPerCapita(stateObj, stateInfo) {
  const pop = stateInfo.population;
  return ((stateObj.value / pop) * 1000).toFixed(2);
}

export const getStateClass = (statesFilter, name) => {
  // no filters so no classes.
  if (statesFilter.length === 0) {
    return '';
  }

  return statesFilter.includes(name) ? 'selected' : 'deselected';
};

export const processStates = (state) => {
  const statesFilter = coalesce(state.query, 'state', []);
  const states = state.map.results.state;
  const stateData = states.map((o) => {
    const stateInfo = coalesce(STATE_DATA, o.name, {
      name: '',
      population: 1,
    });
    o.abbr = o.name;
    o.fullName = stateInfo.name;
    o.perCapita = getPerCapita(o, stateInfo);
    o.className = getStateClass(statesFilter, o.name);
    return o;
  });
  return [stateData];
};

export const mapStateToProps = (state) => {
  const refStateFilters = coalesce(state.query, 'state', []);
  const { isPrintMode, width } = state.view;

  return {
    data: processStates(state),
    dataNormalization: state.query.dataNormalization,
    hasTip: !isPrintMode,
    printClass: isPrintMode ? 'print' : '',
    stateFilters: [...refStateFilters],
    width,
  };
};

export const mapDispatchToProps = (dispatch) => ({
  addState: (selectedState) => {
    sendAnalyticsEvent('State Event: add', selectedState.abbr);
    dispatch(addStateFilter(selectedState));
  },
  removeState: (selectedState) => {
    sendAnalyticsEvent('State Event: remove', selectedState.abbr);
    dispatch(removeStateFilter(selectedState));
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(TileChartMap);
