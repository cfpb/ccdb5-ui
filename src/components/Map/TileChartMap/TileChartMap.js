import './TileChartMap.scss';
import {
  stateFilterAdded,
  stateFilterRemoved,
} from '../../../reducers/filters/filtersSlice';
import { coalesce, sendAnalyticsEvent } from '../../../utils';
import { GEO_NORM_NONE, STATE_DATA } from '../../../constants';
import { useDispatch, useSelector } from 'react-redux';
import { useCallback, useEffect, useMemo } from 'react';
import TileMap from './TileMap';
import {
  selectFiltersDataNormalization,
  selectFiltersState,
} from '../../../reducers/filters/selectors';

import {
  selectViewIsPrintMode,
  selectViewWidth,
} from '../../../reducers/view/selectors';
import { useGetMap } from '../../../api/hooks/useGetMap';

export const TileChartMap = () => {
  const dispatch = useDispatch();
  const dataNormalization = useSelector(selectFiltersDataNormalization);
  const stateFilters = useSelector(selectFiltersState);
  const isPrintMode = useSelector(selectViewIsPrintMode);
  const width = useSelector(selectViewWidth);

  const { data: results } = useGetMap();
  const stateMapResultsState = results?.results.state;

  const data = useMemo(() => {
    return stateMapResultsState
      ? stateMapResultsState.map((state) => {
          const newState = structuredClone(state);
          const stateInfo = coalesce(STATE_DATA, state.name, {
            name: '',
            population: 1,
          });
          newState.abbr = newState.name;
          newState.fullName = stateInfo.name;
          newState.perCapita = getPerCapita(newState, stateInfo);
          return newState;
        })
      : null;
  }, [stateMapResultsState]);

  const hasTip = !isPrintMode;
  const _toggleState = useCallback(
    (event) => {
      // pass in redux dispatch
      // point.fullName
      const { abbr, fullName } = event.point;
      const selectedState = {
        abbr,
        // rename this for consistency
        // chart builder uses fullName
        name: fullName,
      };
      if (stateFilters && stateFilters.includes(abbr)) {
        sendAnalyticsEvent('State Event: remove', selectedState.abbr);
        dispatch(stateFilterRemoved(selectedState));
      } else {
        sendAnalyticsEvent('State Event: add', selectedState.abbr);
        dispatch(stateFilterAdded(selectedState));
      }
    },
    [stateFilters, dispatch],
  );

  const _redrawMap = useCallback(() => {
    const mapElement = document.getElementById('tile-chart-map');
    const mapWidth = isPrintMode ? 650 : mapElement.clientWidth || width;
    if (!data) {
      return;
    }

    const dataSet = updateData(data, dataNormalization, stateFilters);

    const options = {
      el: mapElement,
      data: dataSet,
      isPerCapita: dataNormalization !== GEO_NORM_NONE,
      events: {
        // custom event handlers we can pass on
        click: _toggleState,
      },
      hasTip,
      width: mapWidth,
    };

    options.height = mapWidth * 0.75;

    // eslint-disable-next-line no-unused-vars
    const chart = new TileMap(options);
  }, [
    data,
    dataNormalization,
    hasTip,
    isPrintMode,
    stateFilters,
    width,
    _toggleState,
  ]);

  useEffect(() => {
    _redrawMap();
    return function cleanup() {
      const mapElement = document.getElementById('tile-chart-map');
      if (mapElement) {
        while (mapElement.firstChild) {
          mapElement.firstChild.remove();
        }
      }
    };
  }, [_redrawMap]);

  return (
    <div>
      <div
        id="tile-chart-map"
        className={`cfpb-chart ${isPrintMode ? 'print' : ''}`}
        data-chart-type="tile_map"
        data-testid="tile-chart-map"
      />
    </div>
  );
};

export const getStateClass = (statesFilter, name) => {
  // no filters so no classes.
  if (!statesFilter || statesFilter.length === 0) {
    return '';
  }

  return statesFilter.includes(name) ? 'selected' : 'deselected';
};

/**
 * Helper function to calculate Per Capita value
 *
 * @param {object} stateObj - A state containing abbr and value
 * @param {object} stateInfo - other information about the state
 * @returns {string} the Per 1000 population value
 */
function getPerCapita(stateObj, stateInfo) {
  const pop = stateInfo.population;
  return ((stateObj.value / pop) * 1000).toFixed(2);
}

/**
 * Helper function to get display value of tile based on Normalization.
 *
 * @param {Array} data - Tiles to display.
 * @param {string} dataNormalization - Whether to normalize the data.
 * @param {Array} statesFilter - The currently applied states filter.
 * @returns {object} Data provided to tile map
 */
function updateData(data, dataNormalization, statesFilter) {
  const showDefault = dataNormalization === GEO_NORM_NONE;
  const res = data.map((datum) => ({
    ...datum,
    displayValue: showDefault ? datum.value : datum.perCapita,
    className: getStateClass(statesFilter, datum.name),
  }));

  return res;
}
