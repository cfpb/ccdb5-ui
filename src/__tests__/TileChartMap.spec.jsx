import configureMockStore from 'redux-mock-store'
import { mapStateToProps, TileChartMap } from '../TileChartMap'
import { Provider } from 'react-redux'
import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import thunk from 'redux-thunk'
import { TileMap } from 'cfpb-chart-builder'

// this is how you override and mock an imported constructor
jest.mock( 'cfpb-chart-builder', () => {
  return {
    TileMap: jest.fn().mockImplementation( () => {
      return {}
    } )
  }
} )

function setupSnapshot() {
  const middlewares = [ thunk ]
  const mockStore = configureMockStore( middlewares )
  const store = mockStore( {
    map: {}
  } )

  return renderer.create(
    <Provider store={ store }>
      <TileChartMap/>
    </Provider>
  )
}

describe( 'component: TileChartMap', () => {
  describe( 'initial state', () => {
    it( 'renders without crashing', () => {
      const target = setupSnapshot()
      let tree = target.toJSON()
      expect( tree ).toMatchSnapshot()
    } )
  } )

  describe( 'componentDidUpdate', () => {
    beforeAll( () => {
      const div = document.createElement( 'div' )
      div.setAttribute( 'id', 'tile-chart-map' )
      window.domNode = div
      document.body.appendChild( div )
    } )

    afterAll( () => {
      const div = document.getElementById( 'tile-chart-map' )
      if ( div ) {
        document.body.removeChild( div )
      }
    } )

    it( 'does nothing when no data', () => {
      const target = shallow( <TileChartMap data={ [] }/> )
      target._redrawMap = jest.fn()
      target.setProps( { data: [ [] ] } )
      expect( TileMap ).toHaveBeenCalledTimes( 0 )
    } )

    it( 'skips redraw when the data is the same', () => {
      const target = shallow( <TileChartMap data={ [ [23, 4, 3] ] }/> )
      target._redrawMap = jest.fn()
      target.setProps( { data: [ [23, 4, 3] ] } )
      expect( TileMap ).toHaveBeenCalledTimes( 0 )

    } )

    it( 'trigger a new update when data changes', () => {
      const target = shallow( <TileChartMap data={ [ [23, 4, 3] ] }/> )
      target._redrawMap = jest.fn()
      target.setProps( { data: [ [ 2, 5 ] ] } )
      expect( TileMap ).toHaveBeenCalledTimes( 1 )
    } )
  } )

  describe( 'mapStateToProps', () => {
    it( 'maps state and props', () => {
      const state = {
        map: {
          state: [ 1, 2, 3 ]
        }
      }
      let actual = mapStateToProps( state )
      expect( actual ).toEqual( {
        data: [
          [ 1, 2, 3 ]
        ]
      } )
    } )
  } )


} )
