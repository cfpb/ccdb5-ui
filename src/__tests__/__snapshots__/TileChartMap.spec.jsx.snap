// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: TileChartMap initial state renders without crashing 1`] = `
<div>
  <div
    className="cfpb-chart"
    data-chart-type="tile_map"
    id="tile-chart-map"
  />
</div>
`;
